Set priority for multiple files in one WebAPI request

Piccirello opened the pull request:
This API was never changed during the API revamp some months back

Closes #6259

Chocobo1 commented on src/webui/www/private/scripts/prop-files.js:
```diff
@@ -146,24 +161,24 @@ var createPriorityCombo = function(id, selected_prio) {
-    var opt = new Element("option");
-    opt.set('value', '1');
-    opt.set('html', "QBT_TR(Normal)QBT_TR[CONTEXT=PropListDelegate]");
+    var normal = new Element("option");
```
maybe add a lambda function, so the code will be a bit shorter.

glassez commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -559,12 +559,20 @@ void TorrentsController::filePrioAction()
+    if (torrent && torrent->hasMetadata()) {
+        for (const auto &fileID : fileIDs) {
+            const int id = fileID.toInt(&ok);
+            if (ok)
```
Why invalid file id isn't bad param error? (Unlike invalid prio)

Piccirello replied:
> Just check that they are all correct data type

We have two options: check that they're valid `int`s (which I'm already doing), or check that they're a valid `FilePriority` values.

> P.S. please, don't remove your comments even if you change your mind later. Just say about. Otherwise it's hard to maintain the discussion.

No one removed comments. You replied to the wrong comment thread...

glassez replied:
>No one removed comments. You replied to the wrong comment thread...

Forgive. I just got an email notification, and it was like you answered my question in this thread. But then I didn't find your answer here.

glassez commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -559,12 +559,20 @@ void TorrentsController::filePrioAction()
-    int fileID = params()["id"].toInt();
-    int priority = params()["priority"].toInt();
-    BitTorrent::TorrentHandle *const torrent = BitTorrent::Session::instance()->findTorrent(hash);
+    const QStringList fileIDs {params()["id"].split('|')};
```
I would change param name, but it breaks API compatibility...

Piccirello replied:
I want your thoughts: we add an optional `ids` param that takes priority over the `id` param. Or we can change the name and bump the web api version.

glassez replied:
We shouldn't produce incompatible changes so frequently if possible.
So maybe we should have some way to keep track of these "non-urgent" changes, and then make all the changes at once when we have the required changes?

glassez replied:
Maybe keep the list of postponed changes near the API version declaration, e.g. "When your changes require increasing of minor version you should also apply the following:" and then list of changes.
What do you say? @Chocobo1?

Chocobo1 replied:
>What do you say? @Chocobo1?

I don't think I get your idea.

However I think the commit bumping API version should also include the api code changes, so it is possible to look up by reading git log.

glassez replied:
>I don't think I get your idea.

@Chocobo1, take this case for example.
Meaning of param 'id' was changed. Now it provide list of file IDs. It should cause changing of param name but it breaks API compatibility. It's really bad idea to make incompatible changes too frequently.
We can proceed to use old name for some time and just mark it for future changing.
When we make changes that cannot be made in a compatible way, we can add all pending changes with them.

Chocobo1 replied:
>We can proceed to use old name for some time and just mark it for future changing.

I don't object the idea, however I must say from my experience, revisiting an older code after some time takes much more time for the developer and outcome is usually not good enough quality.
If @Piccirello is interested, we could try it out in this case (experiment).

glassez replied:
>however I must say from my experience, revisiting an older code after some time takes much more time for the developer and outcome is usually not good enough quality.

So I suggest to keep list with postponed changes. We can have some instructions, where, what and how it should be changed. Or we can even have commented changes inplace.

Well, I'm not insisting. I just thought we shouldn't break API too often.

Piccirello replied:
I would be interested, maybe maintaining a separate branch with breaking api changes that can be merged pre-minor version bump.

glassez replied:
>maybe maintaining a separate branch with breaking api changes that can be merged pre-minor version bump.

And have opened PR for it.
Let's try something!

Piccirello commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -559,12 +559,20 @@ void TorrentsController::filePrioAction()
-        torrent->setFilePriority(fileID, priority);
+    BitTorrent::TorrentHandle *const torrent = BitTorrent::Session::instance()->findTorrent(hash);
+    if (torrent && torrent->hasMetadata()) {
+        for (const auto &fileID : fileIDs) {
```
Any benefit to using an explicit type vs `auto`?

Chocobo1 replied:
I believe CODING_GUIDELINES has enough explaination: https://github.com/qbittorrent/qBittorrent/blob/master/CODING_GUIDELINES.md#9-misc

Piccirello replied:
Recommendations for modern C++ are to use `auto` quite liberally. We get no benefit from an explicit type here.

Chocobo1 replied:
>Recommendations for modern C++ are to use auto quite liberally. 

Source?
IMO you're using `auto` a bit too liberal. `auto` is more like a shortcut, not be used like `var` in javascript.

Piccirello replied:
From *Item 5: Prefer auto to explicit type declarations* in *Effective Modern C++* by Scott Meyers:

> Using [*auto*] saves typing, sure, but it also prevents correctness and performance issues that can bedevil manual type declaration ... There are thus several reasons to prefer *auto* over explicit type declarations ... The fact of the matter is that writing types explicitly often does little more than introduce opportunities for subtle errors, either in correctness or efficiency or both.

Chocobo1 replied:
Ah, thank you, however I don't see he mentions anything about readability, a google search gives a handful of debates about readability of using `auto`.

glassez replied:
When we discussed our rules about using `auto` keyword, we wanted to find a balance between faster code writing and readability. So don't abuse it, please.

Piccirello commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -559,12 +559,20 @@ void TorrentsController::filePrioAction()
-    BitTorrent::TorrentHandle *const torrent = BitTorrent::Session::instance()->findTorrent(hash);
+    const QStringList fileIDs {params()["id"].split('|')};
+    bool ok;
+    const int priority = params()["priority"].toInt(&ok);
```
The `FilePriority` enum is defined in `gui/torrentcontentmodelitem.h`, which we can't import into a web controller. Any thoughts on where we should migrate this enum? I would think somewhere inside `base` but I don't know those classes very well.

Piccirello replied:
I think you missed my point. The valid priority values are defined in `gui/torrentcontentmodelitem.h`, which is the list I can check against to see if the value will be accepted.

Piccirello replied:
> offtopic, it seems torrent->setFilePriority only changes 1 prio at a time which is inefficient when you invoke it in a loop.
 
My latest commit modifies `TorrentHandle::setFilePriority()`to take a list of indexes, since the controller code is the only place we call it.

Chocobo1 replied:
>The valid priority values are defined in gui/torrentcontentmodelitem.h, which is the list I can check against to see if the value will be accepted.

That is the gui part, but really who is consuming the priority values? libtorrent, you should check from there.

Piccirello replied:
It's not just for the gui, because we use the same priority values in the webui. The values are defined in gui code because we have very little separation between gui logic and application code, but they really should be moved to base.

libtorrent will accept a priority between 0-7 ([source](https://www.libtorrent.org/reference-Core.html#piece-priority-prioritize-pieces-piece-priorities)); we currently only use 0, 1, 6, and 7. Without the values in `gui/torrentcontentmodelitem.h`, users would be able to enter a value between 2-5.

I'm moving the `prio` namespace to a new file `base/torrentpriority.h`. If we should move it, lmk.

glassez replied:
>That is the gui part, but really who is consuming the priority values? libtorrent, you should check from there.

You shouldn't use libtorrent interface directly outside of abstraction layer (BitTorrent:: types and functions).
>If we should move it, lmk.

Yes. Torrent file priority enum should be declared in BitTorrent namespace. Maybe in torrenthandle.h file (I'm not sure there is separate file should exist).

glassez replied:
>My latest commit modifies `TorrentHandle::setFilePriority()` to take a list of indexes, since the controller code is the only place we call it.

Isn't `prioritizeFiles()` enough?

Piccirello replied:
We still have to do some extra accounting since `prioritizeFiles()` wants a list of priorities for all files in the torrent. But I think that's very doable and will allow us to get rid of `TorrentHandle::setFilePriority()` entirely.

Chocobo1 commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -560,19 +560,25 @@ void TorrentsController::filePrioAction()
+    for (const QString &fileID : fileIDs) {
+        const int id = fileID.toInt(&ok);
+        if (!ok)
+            throw APIError(APIErrorType::BadParams, tr("File ids must be integers"));
```
File IDs ?

Piccirello replied:
Because the user can enter multiple.

Chocobo1 replied:
I mean "ID" in uppercase.

Piccirello replied:
Oh, I misunderstood. Will update

Chocobo1 commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -560,19 +560,25 @@ void TorrentsController::filePrioAction()
+        fileIdsInts << id;
     }
+
+    for (const auto &id : fileIdsInts)
```
`const int id : fileIdsInts`

Piccirello replied:
From the coding guidelines:

`Declarations for which one can gather enough information about the object interface (type) from its name or the usage pattern .. or the right part of the expression nicely fit here.`

glassez replied:
The variable name does not matter. This paragraph means something completely different:
```c++
auto *varName = static_cast<RSS::Feed *>(rssItemPtr);
```
Or:
```c++
auto *session = BitTorrent::Session::instance();
```

Chocobo1 commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -559,12 +559,26 @@ void TorrentsController::filePrioAction()
+    if (!torrent || !torrent->hasMetadata())
+        throw APIError(APIErrorType::BadParams, tr("Torrent hash is not valid"));
+
+    QList<int> fileIdsInts;
```
please keep the code succinct, with this var introduced `QStringList fileIDs` is redundant now.

glassez replied:
fileIDs

glassez commented on src/base/bittorrent/torrenthandle.h:
```diff
@@ -82,6 +82,20 @@ namespace libtorrent
 
 namespace BitTorrent
 {
+    namespace prio
```
This namespace is redundant. It was a very outdated piece of code...
Either declare FilePriority as enum class or move it to TorrentHandle and use more meaningful names: IgnoredFile, NormalPriorityFile, etc. The second way will be easier for you, because it will require less changes to the existing code (enum class will require explicit cast to int).

Piccirello replied:
What's wrong with removing the `prio` namespace and keeping `FilePriority` as an enum (non-class)?

glassez replied:
Nothing wrong for me if we get meaningful names.

Piccirello replied:
To me, ‘BitTorrent::FilePriority::IGNORED’ is descriptive and meaningful, but I take it you don’t agree?

glassez replied:
Yes, it's descriptive and meaningful. I'm OK with it if you'll use camel case: `BitTorrent::FilePriority::Ignored`.
But the question is, how are you going to get those names? Old style enum doesn't produce its own namespace. So you either have to use the old "perverse" way of declaring it in an extra "FilePriority" namespace, or you still have to use an enum class.

Piccirello replied:
I've moved the values into a `FilePriority` class. This also allows us to support an `isValidPriority()` function, which is needed to validate values from the WebAPI.

glassez commented on src/base/bittorrent/torrenthandle.h:
```diff
@@ -82,6 +82,20 @@ namespace libtorrent
+    {
+        enum FilePriority
+        {
+            IGNORED=0,
```
Anyway coding style should be fixed. Add spaces around `=`.

Piccirello commented:
Updated to reflect new column names and rebased.

Before:
<img width="962" alt="screen shot 2018-10-21 at 8 04 28 pm" src="https://user-images.githubusercontent.com/8296030/47274030-ac60a800-d56c-11e8-9567-ac122a0b7895.png">

After:
<img width="959" alt="screen shot 2018-10-21 at 7 49 50 pm" src="https://user-images.githubusercontent.com/8296030/47274031-b1bdf280-d56c-11e8-8421-fd1706000db9.png">

Piccirello commented:
Updated to reflect new column names and rebased.

glassez commented on src/base/bittorrent/torrenthandle.h:
```diff
@@ -155,6 +155,33 @@ namespace BitTorrent
         Error
     };
 
+    class FilePriority
```
Very peculiar use of the class... You just use it as a namespace. I don't think that makes any sense.

Piccirello replied:
So then I'll go back to having an explicit namespace for this. An enum class alone doesn't satisfy my requirements of having an `isValidPriority()` method.

Piccirello replied:
This has been implemented as an enum class now, with `isValidFilePriority()` defined in an anonymous namespace inside `torrentscontroller.cpp`.

glassez commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -559,12 +559,43 @@ void TorrentsController::filePrioAction()
+        throw APIError(APIErrorType::Conflict, tr("Torrent's metadata has not yet downloaded"));
+
+    const int filesCount = torrent->filesCount();
+    QList<int> fileIds;
```
It's better to avoid using of QList unless it's strongly required. Use QVector instead. And reserve needed amount of items before fill it.

Piccirello replied:
Will be removed due to merging the two loops

glassez commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -559,12 +559,43 @@ void TorrentsController::filePrioAction()
+    QVector<int> priorities = torrent->filePriorities();
+    bool priorityChanged = false;
+
+    for (const int &id : fileIds) {
```
Why not to merge this loop with the one above?

Piccirello replied:
It made sense before, but can be refactored now that we use `prioritizeFiles()`.

glassez commented on src/gui/addnewtorrentdialog.cpp:
```diff
@@ -622,18 +622,19 @@ void AddNewTorrentDialog::displayContentTreeMenu(const QPoint &)
-                prio = static_cast<int>(BitTorrent::FilePriority::Ignored);
+                prio = BitTorrent::FilePriority::Ignored;
 
+            int prioInt = static_cast<int>(prio);
```
I wouldn't create this var and just use `static_cast<int>(prio)` inplace.

glassez commented on src/gui/properties/propertieswidget.cpp:
```diff
@@ -620,18 +620,19 @@ void PropertiesWidget::displayFilesListMenu(const QPoint &)
-            prio = static_cast<int>(BitTorrent::FilePriority::Ignored);
+            prio = BitTorrent::FilePriority::Ignored;
 
+        int prioInt = static_cast<int>(prio);
```
The same as above.

glassez commented on src/gui/properties/proplistdelegate.cpp:
```diff
@@ -111,16 +112,16 @@ void PropListDelegate::paint(QPainter *painter, const QStyleOptionViewItem &opti
     case PRIORITY: {
             QString text = "";
             switch (index.data().toInt()) {
```
```c++
switch (static_cast<BitTorrent::FilePriority>(index.data().toInt())) {
case BitTorrent::FilePriority::Mixed:
// ... and so on
}
```

glassez commented on src/gui/properties/proplistdelegate.cpp:
```diff
@@ -155,13 +156,13 @@ void PropListDelegate::setEditorData(QWidget *editor, const QModelIndex &index)
     QComboBox *combobox = static_cast<QComboBox *>(editor);
     // Set combobox index
     switch (index.data().toInt()) {
```
The same as above.

glassez commented on src/gui/torrentcontentmodelitem.cpp:
```diff
@@ -38,7 +39,7 @@ TorrentContentModelItem::TorrentContentModelItem(TorrentContentModelFolder *pare
     , m_size(0)
     , m_remaining(0)
-    , m_priority(prio::NORMAL)
+    , m_priority(static_cast<int>(BitTorrent::FilePriority::Normal))
```
Seems like it would be better to have `m_priority` and all related methods (here and in child classes) of type FilePriority.
I don't insist that you change it in this PR if it's too much for you.

xnoreq commented:
Why would you concatenate ids with pipes into a long string just to later split them again?!

Why not do the proper thing and use a field "ids" that has an array of integer IDs as value?

So instead of `"id": "134|140|500|...."` you have `"id": [134, 140, 500, ...]`.
Instead of `"id": ids.join("|")` you have `"ids": ids`.
Instead of doing the split server-side you iterate over the array of integers directly.

Piccirello commented:
@xnoreq This is currently done to maintain consistency with all our existing APIs. In the future these should all be cleaned up and switched to the "correct" way.

Chocobo1 commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -124,6 +124,20 @@ namespace
         }
     }
+
+    bool isValidFilePriority(BitTorrent::FilePriority priority)
```
`bool isValidFilePriority(const BitTorrent::FilePriority priority)`

glassez commented on src/gui/torrentcontentmodelitem.h:
```diff
@@ -32,17 +32,7 @@
-        MIXED=-1
-    };
-}
+#include "base/bittorrent/torrenthandle.h"
```
Seems like **FilePriority** should be declared in its own header...

Piccirello replied:
Do we want a new header file just for this one enum class? Seems it might be overkill, but I don't feel strongly.

glassez replied:
Overkill is include entire "torrenthandle.h" when we need only FilePriority class. @Chocobo1, what do you say? If it's no problem feel free to merge it.

Chocobo1 replied:
This idea actually has popped into my mind, I think there should be a file that has all these similar enum/struct defines (and their helper functions). Maybe "bitorrent/enums.{cpp, h}"?

glassez replied:
>I think there should be a file that has all these similar enum/struct defines (and their helper functions). Maybe "bitorrent/enums.{cpp, h}"?

I would not combine different types in one header (unless they are very likely used adjacently).

Piccirello replied:
I'll move the enum class to its own header file.

Piccirello replied:
Move is complete

glassez commented on src/base/bittorrent/filepriority.h:
```diff
@@ -0,0 +1,44 @@
+ * exception statement from your version.
+ */
+
+#ifndef BITTORRENT_FILEPRIORITY_H
```
Please replace it with `#pragma once`

glassez commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -559,12 +574,38 @@ void TorrentsController::filePrioAction()
+        if (!ok)
+            throw APIError(APIErrorType::BadParams, tr("File IDs must be integers"));
+        if ((id < 0) || (id >= filesCount))
+            throw APIError(APIErrorType::NotFound, tr("File ID is not valid"));
```
Conflict

Chocobo1 commented on src/webui/api/torrentscontroller.cpp:
```diff
@@ -124,6 +125,20 @@ namespace
         }
     }
+
+    bool isValidFilePriority(const BitTorrent::FilePriority priority)
```
I think this should be in filepriority.cpp.

Piccirello replied:
Agreed it should move to filepriority. But surely it can live inside the header file, no?

Chocobo1 replied:
The function will be inlined if placed in header, but it is only used in 1 place so OK.

glassez replied:
Coding Guidelines:
>Method definitions aren't allowed in header files

> but it is only used in 1 place so OK

This may change.

Piccirello replied:
I'm having some trouble with this. I've created a `filepriority.cpp` with the following contents, and included it in `base.pri` and the appropriate `CMakeLists.txt`

```
#include "filepriority.h"

namespace BitTorrent
{
    static bool isValidFilePriority(const BitTorrent::FilePriority priority)
    {
        switch (priority) {
        case BitTorrent::FilePriority::Ignored:
        case BitTorrent::FilePriority::Normal:
        case BitTorrent::FilePriority::High:
        case BitTorrent::FilePriority::Maximum:
        case BitTorrent::FilePriority::Mixed:
            return true;
        default:
            return false;
        }
    }
}

```

but I get this compiler error:
> webui/api/torrentscontroller.cpp:570:10: error: ‘isValidFilePriority’ is not a member of ‘BitTorrent’

glassez replied:
1. You should declare it in "filepriority.h" and then define in "filepriority.cpp"
2. You have not use `static` with it. `static` declares functions with **internal linkage** (i.e. it can be referred to from all scopes in the *current translation unit* only). Actually anonymous namespaces are used for the same purposes currently. But you do want to use it outside of "filepriority.cpp"! So you don't need it to be internally linked.

Piccirello replied:
Removing the `static` keyword fixed it

Chocobo1 commented on src/base/bittorrent/filepriority.h:
```diff
@@ -0,0 +1,43 @@
+        Mixed = -1
+    };
+
+    bool isValidFilePriority(const BitTorrent::FilePriority priority);
```
@Piccirello 
Please remove the `const` here (header file) and we can merge.

Piccirello replied:
@Chocobo1 Why just the header file? Possibly some C++ specifics I don't know.

Piccirello replied:
@glassez how is this example any different than [this](https://github.com/qbittorrent/qBittorrent/pull/9799#discussion_r230601668)? The recommendations on passing by const value seem to be inconsistent.

glassez replied:
>how is this example any different than this?

This has been mentioned many times (including in conversations with you).
It may make sense in function definition, but it is completely meaningless in the function declaration (moreover, it is misleading there).

Chocobo1 replied:
@Piccirello 
>It may make sense in function definition

The usual advantages of using `const`.

>but it is completely meaningless in the function declaration 

Because it is a local variable, `const` or not is not relevant to the users of the the function, so we leave it out.

Piccirello replied:
Thanks for the explainer. I've pushed the change.

Chocobo1 commented:
Thank you!
